### PR TITLE
only allow usage of whitelisted secret

### PR DESCRIPTION
**Pull request by honnix**
# Hey, I just made a Pull Request!

## Description

Only allow usage of whitelisted secrets, otherwise workflow creation/update/execution will fail.

## Checklist for PR author(s)

- [x] Changes are covered by unit test
- [x] All tests pass
- [x] Code coverage check passes
- [ ] Error handling is tested
- [ ] Errors are handled at the appropriate layer
- [ ] Errors that cannot be handled where they occur are propagated
- [ ] (optional) Changes are covered by system test
- [ ] Relevant documentation updated
- [x] This PR has NO breaking change to public API
- [ ] This PR has breaking change to public API and it is documented

## Checklist for PR reviewer(s)

- [x] This PR has been incorporated in release note for the coming version
- [ ] Risky changes introduced by this PR have been all considered

### CODE REVIEW

File: styx-common/src/main/java/com/spotify/styx/util/WorkflowValidator.java [hunk 1]
```diff
@@ -46,12 +48,15 @@
...
   private final Duration maybeMaxRunningTimeout;
+  private final Set<String> secretWhitelist;
 
-  private WorkflowValidator(DockerImageValidator dockerImageValidator, Duration maybeMaxRunningTimeout) {
+  private WorkflowValidator(DockerImageValidator dockerImageValidator, Duration maybeMaxRunningTimeout,
+                            Set<String> secretWhitelist) {
     Preconditions.checkArgument(maybeMaxRunningTimeout == null || !maybeMaxRunningTimeout.isNegative(),
         "Max Running timeout should be positive");
-    this.dockerImageValidator = dockerImageValidator;
+    this.dockerImageValidator = Objects.requireNonNull(dockerImageValidator);
```
Comment (narape) [inline styx-common/src/main/java/com/spotify/styx/util/WorkflowValidator.java]: Can you add `requireNonNull` also for `secretWhitelist` to make it consistent and more error proof?

File: styx-common/src/main/java/com/spotify/styx/util/WorkflowValidator.java [hunk 2]
```diff
@@ -133,6 +138,12 @@ public static Builder newBuilder(DockerImageValidator dockerImageValidator) {
       }
     });
 
+    cfg.secret().ifPresent(secret -> {
+      if (secretWhitelist != null && !secretWhitelist.contains(secret.name())) {
```
Comment (narape) [inline styx-common/src/main/java/com/spotify/styx/util/WorkflowValidator.java]: I would rather initialize `secretWhitelist` to `Set.of()` (the empty set) in the `Builder` class so we can simplify this line to be just 

```java
if (!secretWhitelist.contains(secret.name())) {
```

File: styx-common/src/test/java/com/spotify/styx/util/WorkflowValidatorTest.java [hunk 3]
```diff
@@ -247,6 +248,30 @@ public void shouldNotAllowComponentWithHash(String component) {
...
+    WorkflowValidator sut = WorkflowValidator.newBuilder(dockerImageValidator)
+        .withSecretWhitelist(Set.of(FULL_WORKFLOW_CONFIGURATION.secret().get().name()))
+        .build();
+
+    final List<String> errors = sut.validateWorkflow(
+        Workflow.create("test", FULL_WORKFLOW_CONFIGURATION));
+
+    assertThat(errors, empty());
+  }
+
```
Comment (narape) [inline styx-common/src/test/java/com/spotify/styx/util/WorkflowValidatorTest.java]: Can you add also a test named like `shouldPassUsageForEmptyWhitelist` to verify that we wont fail if there is no white list?

File: styx-common/src/main/java/com/spotify/styx/util/WorkflowValidator.java (hunk 2 above)
Comment (honnix) [reply styx-common/src/main/java/com/spotify/styx/util/WorkflowValidator.java]: Unfortunately this validator is used by CLIMain.java where we can not apply any whitelist check. So `null` means skipping whitelist check.

File: styx-common/src/main/java/com/spotify/styx/util/WorkflowValidator.java (hunk 1 above)
Comment (honnix) [reply styx-common/src/main/java/com/spotify/styx/util/WorkflowValidator.java]: Unfortunately this validator is used by CLIMain.java where we can not apply any whitelist check. So `null` means skipping whitelist check.

File: styx-common/src/main/java/com/spotify/styx/util/WorkflowValidator.java (hunk 2 above)
Comment (honnix) [reply styx-common/src/main/java/com/spotify/styx/util/WorkflowValidator.java]: Unless we say `Set.of("*")` is a wildcard.

File: styx-common/src/test/java/com/spotify/styx/util/WorkflowValidatorTest.java (hunk 3 above)
Comment (honnix) [reply styx-common/src/test/java/com/spotify/styx/util/WorkflowValidatorTest.java]: I think this is covered by all the other test cases where we use `sut`.

File: styx-common/src/main/java/com/spotify/styx/util/WorkflowValidator.java (hunk 1 above)
Comment (narape) [reply styx-common/src/main/java/com/spotify/styx/util/WorkflowValidator.java]: Yes, I see

File: styx-common/src/main/java/com/spotify/styx/util/WorkflowValidator.java (hunk 2 above)
Comment (narape) [reply styx-common/src/main/java/com/spotify/styx/util/WorkflowValidator.java]: Wildcard will make it more messy. As we discussed earlier the problem with the nulls are that we are using the same class in the cli and the service. What we need is actually split this class and make the validators composable but that is for another PR

File: styx-common/src/test/java/com/spotify/styx/util/WorkflowValidatorTest.java (hunk 3 above)
Comment (narape) [reply styx-common/src/test/java/com/spotify/styx/util/WorkflowValidatorTest.java]: Agree

**Pull request merged**